### ISSUE TITLE
QueryResult loading not 'true' when changing variables (in QueryOptions)

### ISSUE THREAD

**Issue by lemonbuzz** [zino-app/graphql-flutter#351]
So basically, when I fetch the first time(first build), QueryResult 'loading' property returns 'true' which is expected. However when I change the parameters of my query(dynamically, which triggers a rebuild), the loading property stays 'false', but I still end up getting the new data.

Is there something wrong with my approach?

**Comment by mainawycliffe** [comment]
Can you try and call [setState](https://api.flutter.dev/flutter/widgets/State/setState.html) after changing the variables. Please note, this will cause you entire stateful widget to rebuild, but for your case it might have the desired effect. 

The other option, is to wait until we have finalized the `fetchMore` feature, this would be a good use case for it.

**Comment by lemonbuzz** [comment]
Yeah setState is already in place! Everything is fine on the rebuild, the query gets called but loading is false

**Comment by mainawycliffe** [comment]
Are you using caching? What is your fetchPolicy?

**Comment by lemonbuzz** [comment]
Default fetchPolicy. For cache its InMemoryCache

**Comment by mainawycliffe** [comment]
The default one is cache first, which first loads data from the cache, and then looks for fresh data from the network. When new data is received, it replaces the stale data. Set the fetch policy to Network Only and see if it works as you want it to.

**Comment by lemonbuzz** [comment]
Yeah I already tried all the different fetch policies. Not sure what is going on

**Comment by mainawycliffe** [comment]
Sorry for the late response, I would like to have a closer look at this. can you share a code to replicate, i want to have a closer look at this?

**Comment by lemonbuzz** [comment]
Hey thanks for the follow up! I cannot share the code, but I was able to reproduce it easily with the example provided by the graphql-flutter repo. I simply use the counter to increment the query variable `nRepositories`, but loading is always false ( except for the first initial build ). The query results eventually update (according to nRepositories) but there is no way to know if it's loading. 

```
import 'package:flutter/material.dart';
import 'package:graphql_flutter/graphql_flutter.dart';

void main() => runApp(MyApp());

String readRepositories = """
  query ReadRepositories(\$nRepositories: Int!) {
    viewer {
      repositories(last: \$nRepositories) {
        nodes {
          id
          name
          viewerHasStarred
        }
      }
    }
  }
""";

final HttpLink httpLink = HttpLink(
  uri: 'https://api.github.com/graphql',
);

final AuthLink authLink = AuthLink(
  getToken: () async => 'Bearer <TOKEN>',
);

final Link link = authLink.concat(httpLink as Link);

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return GraphQLProvider(
      client: ValueNotifier(
        GraphQLClient(
          cache: InMemoryCache(),
          link: link,
        ),
      ),
      child: MaterialApp(
        title: 'Flutter Demo',
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: MyHomePage(title: 'Flutter Demo Home Page'),
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Query(
        options: QueryOptions(
          document: readRepositories,
          variables: {
            'nRepositories': _counter,
          },
          pollInterval: 10,
        ),
        builder: (QueryResult result, {VoidCallback refetch}) {
          if (result.errors != null) {
            return Text(result.errors.toString());
          }
          print(result.loading);
          if (result.loading) {
            return Text('Loading');
          }
          List repositories = result.data['viewer']['repositories']['nodes'];

          return ListView.builder(
              itemCount: repositories.length,
              itemBuilder: (context, index) {
                final repository = repositories[index];

                return Text(repository['name']);
              });
        },
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ),
    );
  }
}
```

**Comment by mainawycliffe** [comment]
This will do thanks. Let me investigate and see what the issue is

**Comment by mainawycliffe** [comment]
I see where the problem is, it return the previous data while fetching the new data. Am not sure the best way to solve this is. In such cases where the Query Options change, we could return the stale data and also set loading to true, so one can decide whether to show stale data or show a loading indicator or both (I am in favor of this). The other option is to discard the old data once the Query option has change and set loading to true, this would only give the option to show a loading indicator.

**Comment by lemonbuzz** [comment]
Hey thanks for the investigation! I don't have much time for this project anymore, so if you can point me to a quick fix I can implement myself in the library, that would be cool!

As for the actual fix, I think Apollo clears the data when the query parameters changes and loading is set to true. Your first solution is more flexible but coming from Apollo it might be off putting.

**Comment by martinek** [comment]
@mainawycliffe When options change, whole new query should be created since with new variables, the query itself is different. Data do not need to be removed from cache in this case. The data is still valid as cached response for former variables. The result.data should return value for the new query though.

The simple use case for this is if you have widget that takes some item id for example. The widget might load item data. When the id is changed, widget is rebuilt and data for new item should be loaded with loading and all. The old data is not valid for new id, so it should be present anywhere. Then if id change back to previous one, the previous data can be used if returning data from cache is allowed.

There is only one specific case when this might be not true and that is pagination / fetchMore. I think that case (pagination) should be handled on its own as it involves cache manipulation and such.

**Comment by mainawycliffe** [comment]
@martinek Yea, i agree with what you just said. I just wasn't sure this was the way to go, and since a fetchMore/pagination feature is also in the pipeline, its wise to go ahead and clear everything when the variables are build. 

@lemonbuzz from my latest my investigation, the intended function seems to be to clear data, as you have both stated, for some reason, on rebuild, the stream is not getting closed on rebuild, and instead staying around, after rebuild. I will do some more investigation over the weekend, and create a PR.

**Comment by lemonbuzz** [comment]
Thank you! Looking forward to trying it out :)

**Comment by lemonbuzz** [comment]
Any news on this :)?

**Comment by mainawycliffe** [comment]
@lemonbuzz not yet, haven't had any success into identifying where the issue is. Any ideas @micimize?

**Comment by micimize** [comment]
closed by #394
